addon/-private/external/generator-state.js

```diff
@@ -0,0 +1,43 @@
+export class GeneratorStepResult {
+  constructor(value, done, errored) {
+    this.value = value;
+    this.done = done;
+    this.errored = errored;
+  }
+}
+
+export class GeneratorState {
+  constructor(generatorFactory) {
+    this.done = false;
+    this.generatorFactory = generatorFactory;
+    this.iterator = null;
+  }
+
+  step(resolvedValue, iteratorMethod) {
+    try {
+      let iterator = this.getIterator();
+      let { value, done } = iterator[iteratorMethod](resolvedValue);
+
+      if (done) {
+        return this.finalize(value, false);
+      } else {
+        return new GeneratorStepResult(value, false, false);
+      }
+    } catch(e) {
+      return this.finalize(e, true);
+    } 
+  }
+
+  getIterator() {
+    if (!this.iterator && !this.done) {
+      this.iterator = this.generatorFactory();
+    }
+    return this.iterator;
+  }
```

Comment:
Are these _iterators_ different then ES6 iterators (`Symbol.iterator`)?

Reply:
`this.iterator` is a generator function, so it would also be usable with `Symbol.iterator`. However, `GeneratorState` is not as-of-yet accessed this way and is more of a state wrapper/manager around an iterator than an iterator itself.